generated/nidaqmx/_grpc_interpreter.py

```diff
@@ -2579,24 +2594,84 @@ def read_raw(self, task, num_samps_per_chan, timeout, read_array):
+        except grpc.RpcError as rpc_error:
+            self._handle_rpc_error(rpc_error)
+
+
+        def done_event_thread():
+            try:
+                for response in response_stream:
+                    callback_function(task, response.status, callback_data)
+            except grpc.RpcError as e:
+                if e.code() == grpc.StatusCode.CANCELLED:
+                    return
+                raise
+            except errors.RpcError as e:
+                if e.rpc_code == grpc.StatusCode.CANCELLED:
+                    return
+                raise
+
+        thread = threading.Thread(target=done_event_thread, name="NI-DAQmx Done Event Thread")
+        thread.start()
+        return GrpcEventHandler(response_stream, thread)
+
 
     def register_every_n_samples_event(
             self, task, every_n_samples_event_type, n_samples, options,
             callback_function, callback_data):
-        response = self._invoke(
-            self._client.RegisterEveryNSamplesEvent,
-            grpc_types.RegisterEveryNSamplesEventRequest(
-                task=task,
-                every_n_samples_event_type_raw=every_n_samples_event_type,
-                n_samples=n_samples, options=options,
-                callback_function=callback_function,
-                callback_data=callback_data))
+        if callback_function is not None:
+            return self._register_every_n_samples_event(task, every_n_samples_event_type, n_samples, options, callback_function, callback_data)
+        else:
+            raise NotImplementedError("The NI gRPC Device Server does not support event unregistration.")
+
+    def _register_every_n_samples_event(
+            self, task, every_n_samples_event_type, n_samples, options,
+            callback_function, callback_data):
+        try:
+            response_stream = self._client.RegisterEveryNSamplesEvent(
+                grpc_types.RegisterEveryNSamplesEventRequest(
+                    task=task,
+                    every_n_samples_event_type_raw=every_n_samples_event_type,
+                    n_samples=n_samples))
+
+            # Wait for initial metadata to ensure the registration is complete.
+            _ = response_stream.initial_metadata()
+        except grpc.RpcError as rpc_error:
+            self._handle_rpc_error(rpc_error)
+
+        def every_n_samples_event_thread():
+            try:
+                for response in response_stream:
+                    callback_function(
+                        task,
+                        response.every_n_samples_event_type_raw,
+                        response.n_samples,
+                        callback_data)
+            except grpc.RpcError as e:
+                if e.code() == grpc.StatusCode.CANCELLED:
+                    return
+                raise
+            except errors.RpcError as e:
+                if e.rpc_code == grpc.StatusCode.CANCELLED:
+                    return
+                raise
```

Comment:
This is identical to `done_event_thread` except for the args to the callback_function. Making it generic is pretty easy except that you pull different things out of the `response`. Once you generalize that, then its probably not all that useful to share anymore... ah well

Reply:
This duplication is also in https://github.com/ni/nidaqmx-python/pull/343#discussion_r1194248172

I posted some suggestions for refactoring the exception handling:

> Consider adding a private helper function to `_grpc_interpreter.py.mako` like `_is_canceled(ex: Exception) -> bool` or `_get_rpc_code(ex: Exception) -> Optional[grpc.StatusCode]`. I included the type annotations because they communicate what the function does.

It's also possible to write a helper function that takes a callable with no arguments, or a helper function that takes a callable that takes a response.